### DISCUSSION TITLE
Getter & setter functions - allow extra "optional" parameters (i.e. parameters with default assignments)

### DISCUSSION THREAD

**Discussion by jon-heard** [godotengine/godot-proposals#6107] [Scripting]
### Background
Currently, the getter / setter system requires that a getter function takes exactly 0 parameters and that a setter function takes exactly 1 parameter.  If this is not adhered to then Godot errors with: `Function "X" cannot be used as setter because of its signature.`.

If a getter or setter logically could or should take extra parameters, then the current solution is to write an alternative function with the same purpose, but taking those extra parameters.

### Example - a solution that work with the current system
```
var length: float : get = get_length

func get_length() -> float:
    return length

func get_length_of_type(type: DistanceType = DistanceType.texture_pixels) -> void:
    match type:
        DistanceType.scaled:
            return length * scale
        DistanceType.screen:
            return length * scale * zoom
        _:
            return length
```

### Proposal
My suggestion is to accept getter and setter functions that have more than the required number of parameters as long as the extra parameters are optional (i.e. they include a default assignment).

### The benefit
If a getter / setter function logically could or should take extra parameters, then this proposal would allow it to, as long as the extra parameters have defaults that the getter / setter system can use.  The advantage of this proposal is for more straightforward logic as it allows a having a single function for a single purpose rather than two functions for essentially the same purpose.

### Example - a solution that takes advantage of this proposal
```
var length: float : get = get_length

func get_length(type: DistanceType = DistanceType.texture_pixels) -> void:
    match type:
        DistanceType.scaled:
            return length * scale
        DistanceType.screen:
            return length * scale * zoom
        _:
            return length
```

**Comment by jon-heard** [comment]
I have a few viewers responding with confusion faces.  I'm curious as to where the confusion lies.  Perhaps this is a polite way of saying that they don't think that this is a good idea, or a valuable modification?  It'd be great to get some clarification so that I know how to proceed.

**Reply by dalexeev** [reply]
I'm not strictly against it, I just have doubts where it can be used.

Now you can work around it like this:

```gdscript
var my_var:
    get: return get_my_var()
    set(value): set_my_var(value)

func get_my_var(a = 1, b = 2):
    ...

func set_my_var(value, a = 1, b = 2):
    ...
```

**Reply by jon-heard** [reply]
It comes up now and again.  One recent example:

I have a boolean "faded_black" variable to show/hide a ColorRect to blank the screen.  The ColorRect has fading animations, played via the setter.  However, sometimes I want to fade slowly (during the title) or fast (during some cutscenes).  Currently, I have a "set_faded_black" function (the setter) and a "set_faded_black_custom_speed" function, which does the same thing, but with an extra "speed" parameter.

I think the technique you demonstrated is an improvement over the double-function technique I've been using.  I'll adopt it.

I recognize that what I'm proposing has servicable workarounds.  It's really just a convenience, I suppose.

**Reply by jon-heard** [reply]
Thanks for replying, btw.

**Comment by dalexeev** [comment]
See godotengine/godot#74144.

**Comment by dalexeev** [comment]
Let's close, since the behavior was identified as a bug (https://github.com/godotengine/godot/issues/74144). Thanks for the contribution nonetheless!
